flash/data/data_source.py

```diff
@@ -0,0 +1,263 @@
+from flash.data.data_pipeline import DataPipeline
+from flash.data.process import Preprocess, ProcessState
+from flash.data.utils import _STAGES_PREFIX, CurrentRunningStageFuncContext
+
+
+def has_len(data: Union[Sequence[Any], Iterable[Any]]) -> bool:
+    try:
+        len(data)
+        return True
+    except (TypeError, NotImplementedError):
+        return False
+
+
+@dataclass(unsafe_hash=True, frozen=True)
+class LabelsState(ProcessState):
+
+    labels: Optional[Sequence[str]]
+
+
+class MockDataset:
+
+    def __init__(self):
+        self.metadata = {}
+
+    def __setattr__(self, key, value):
+        if key != 'metadata':
+            self.metadata[key] = value
+        else:
+            object.__setattr__(self, key, value)
+
+
+class DataSource(ABC):
+
+    def __init__(
+        self,
+        train_data: Optional[Any] = None,
+        val_data: Optional[Any] = None,
+        test_data: Optional[Any] = None,
+        predict_data: Optional[Any] = None,
+    ):
+        super().__init__()
+
+        self.train_data = train_data
+        self.val_data = val_data
+        self.test_data = test_data
+        self.predict_data = predict_data
+
+        self._preprocess: Optional[Preprocess] = None
+
+    @property
+    def preprocess(self) -> Optional[Preprocess]:
+        return self._preprocess
+
+    @abstractmethod
+    def load_data(
+        self,
+        data: Any,
+        dataset: Optional[Any] = None
+    ) -> Iterable[Mapping[str, Any]]:  # TODO: decide what type this should be
+        """Loads entire data from Dataset. The input ``data`` can be anything, but you need to return a Mapping.
+
+        Example::
+
+            # data: "."
+            # output: [("./cat/1.png", 1), ..., ("./dog/10.png", 0)]
+
+            output: Mapping = load_data(data)
+
+        """
+
+    @abstractmethod
+    def load_sample(self, sample: Mapping[str, Any], dataset: Optional[Any] = None) -> Any:
+        """Loads single sample from dataset"""
+
+    def to_datasets(self, data_pipeline: DataPipeline) -> Tuple[Optional[BaseAutoDataset], ...]:
+        # attach preprocess
+        self._preprocess = data_pipeline._preprocess_pipeline
+
+        train_dataset = self._generate_dataset_if_possible(RunningStage.TRAINING, data_pipeline)
+        val_dataset = self._generate_dataset_if_possible(RunningStage.VALIDATING, data_pipeline)
+        test_dataset = self._generate_dataset_if_possible(RunningStage.TESTING, data_pipeline)
+        predict_dataset = self._generate_dataset_if_possible(RunningStage.PREDICTING, data_pipeline)
+        return train_dataset, val_dataset, test_dataset, predict_dataset
+
+    def _generate_dataset_if_possible(
+        self,
+        running_stage: RunningStage,
+        data_pipeline: DataPipeline,
+    ) -> Optional[Union[AutoDataset, IterableAutoDataset]]:
+        data = getattr(self, f"{_STAGES_PREFIX[running_stage]}_data", None)
```

Comment:
Let's support prefix there for `predict_load_data` there too.

Reply:
Done